### DISCUSSION TITLE
G-Code Viewer Color Mode "Feature" not rendering after selecting one of the other options.

### DISCUSSION THREAD

**Discussion by Giterdun2** [xblax/flashforge_ad5m_klipper_mod#142] [General]
Watching a build printing, I decided to check some of the options in the G-Code Viewer.. The default Color Mode was Feature, which was initially working.. I selected Extruder, and that worked ok, and then Feed Rate, which also worked ok. Then when I re-selected Feature, it wouldn't render, and the rendering window was not closeable with the X. Selecting one of the other 2 options the rendering worked ok again. Changing the Render Quality made no difference.  After printing finished, selecting a local file to render did the same thing.  Anyone else play around with this?

**Comment by consp** [comment]
Is this in fluidd or mainsail?

**Comment by Giterdun2** [comment]
> Is this in fluidd or mainsail?
This is in Mainsail. Sry.. I didn't think to try it in fluidd..

**Reply by consp** [reply]
No problem, but it's easier to check if you know where the issue originated :)

**Comment by xblax** [comment]
Maybe it's this: https://github.com/mainsail-crew/mainsail/pull/1755

We can check if we can update to Mainsail v2.10.0 Release, that includes the fix:
https://github.com/mainsail-crew/mainsail/releases/tag/v2.10.0

Fluidd also has new point release:
https://github.com/fluidd-core/fluidd/releases/tag/v1.29.1

**Reply by Giterdun2** [reply]
Thanks..  I've only ever installed the Klipper Mod that included Mainsail & fluidd.. I've downloaded both updates, but no idea what to do with them.. lol..  Please let me know if we can update without breaking something.. ;)  Then how would one go about updating these browsers on the printer?  Or would these updates just be included in a new Klipper Mod point update?

**Comment by Giterdun2** [comment]
> No problem, but it's easier to check if you know where the issue originated :)

In fluidd, it seems that it only gives the single top view of the file. There's no option that I can find to switch between Extruder, Feature, like in Mainsail.. In any case, since this happened, I've been unable to use "Feature" in the G-Code Viewer in Mainsail. I tried to use kiauh to update Mainsail, but can't seem to get it to work, altho it shows that it's there..  
![kiauh](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/165448658/7ff3b153-7e4b-4fcd-95af-e38f5b4cf0da)

**Reply by consp** [reply]
kiauth is not installed since it will install a (highly likely) incompatible version of the software. While mainsail/fluidd are mostly original, klipper has been slightly modified (and hold back) to stay compatible with the printer. The only way to update the mod is to install the new version via usb (v0.05 is in the works) at this point.

Also: kiauth requires a debian installation with systemd and all kinds of fancy features not supported by the mod (or the system it's running on due to hardware constraints), it assumes a raspberry pi which is an order of magnitude more powerful than what's available.

**Comment by Giterdun2** [comment]
> kiauth is not installed since it will install a (highly likely) incompatible version of the software. While mainsail/fluidd are mostly original, klipper has been slightly modified (and hold back) to stay compatible with the printer. The only way to update the mod is to install the new version via usb (v0.05 is in the works) at this point.
> 
> Also: kiauth requires a debian installation with systemd and all kinds of fancy features not supported by the mod (or the system it's running on due to hardware constraints), it assumes a raspberry pi which is an order of magnitude more powerful than what's available.

Ok thanks very much for the info.. I wasn't aware of the incompatibility, and I never saw anything about it.. Much appreciated. Kinda weird tho as it appeared like it installed in the screenshot. It also shows the folder so not sure what I'm actually looking at there.... I used to dabble a bit in linux years ago, but it's been a while.... lol... By "update the mod" are you referring to the Klipper Mod? I'm using v0.04 now.. I've uninstalled and reinstalled it a couple times now for other reasons..  Is there a way to just update Mainsail on KlipperMod v0.04? or do you know if the Mainsail v2.10.0 version will be included in v0.05 update?

**Reply by consp** [reply]
> By "update the mod" are you referring to the Klipper Mod?

Yes, public release version is v0.04. No need to run separate updates on mainsail or other software. 

> Mainsail v2.10.0 version will be included in v0.05 update

@xblax might know more about this :) Currently it's v2.9.1. Might have missed it somewhere, is there a specific reason to update to that version?

If you have to update, you can login to the system, download the mainsail update and overwrite the version on the system. See the docs dir for how to gain ssh access.

**Reply by xblax** [reply]
Yes it will, be in it's already on a branch. I just need to finish everything, write some documentation and make the release. Hopefully I get it done this weekend.

**Reply by Giterdun2** [reply]
Ok Thanks a lot.. just updated to v0.05.. The rendering with the "Feature option appears to be working now, but when tracking the current file, it shows the full completed image instead of where the build is at that time..  Also, Extruder and Feed Rate do the same except it changes color.. see pic... 
![rendering](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/165448658/d638d676-5f6c-4e95-93ce-22c170c7cbea)
